src/granite/columns.cr

```diff
@@ -32,66 +32,26 @@ module Granite::Columns
-            {% end %}
-        {% end %}
-        end
-      end
-    {% end %}
-  end
-
   # Defines a column *decl* with the given *options*.
   macro column(decl, **options)
     {% type = decl.type %}
-    {% not_nilable_type = type.is_a?(Path) ? type.resolve : (type.is_a?(Union) ? type.types.reject(&.resolve.nilable?).first : (type.is_a?(Generic) ? type.resolve : type)) %}
-
-    # Raise an exception if the delc type has more than 2 union types or if it has 2 types without nil
-    # This prevents having a column typed to String | Int32 etc.
     {% if type.is_a?(Union) && (type.types.size > 2 || (type.types.size == 2 && !type.types.any?(&.resolve.nilable?))) %}
       {% raise "The column #{@type.name}##{decl.var} cannot consist of a Union with a type other than `Nil`." %}
     {% end %}
 
+    {% nilable = type.resolve.nilable? %}
     {% column_type = (options[:column_type] && !options[:column_type].nil?) ? options[:column_type] : nil %}
     {% converter = (options[:converter] && !options[:converter].nil?) ? options[:converter] : nil %}
     {% primary = (options[:primary] && !options[:primary].nil?) ? options[:primary] : false %}
     {% auto = (options[:auto] && !options[:auto].nil?) ? options[:auto] : false %}
     {% auto = (!options || (options && options[:auto] == nil)) && primary %}
 
-    {% nilable = (type.is_a?(Path) ? type.resolve.nilable? : (type.is_a?(Union) ? type.types.any?(&.resolve.nilable?) : (type.is_a?(Generic) ? type.resolve.nilable? : type.nilable?))) %}
+    {% if primary && !nilable %}
+      {% raise "Primary key of #{@type} must be nilable" %}
+    {% end %}
 
     @[Granite::Column(column_type: {{column_type}}, converter: {{converter}}, auto: {{auto}}, primary: {{primary}}, nilable: {{nilable}})]
-    @{{decl.var}} : {{decl.type}}? {% unless decl.value.is_a? Nop %} = {{decl.value}} {% end %}
-
-    {% if nilable || primary %}
-      def {{decl.var.id}}=(@{{decl.var.id}} : {{not_nilable_type}}?); end
-
-      def {{decl.var.id}} : {{not_nilable_type}}?
-        @{{decl.var}}
-      end
-
-      def {{decl.var.id}}! : {{not_nilable_type}}
-        raise NilAssertionError.new {{@type.name.stringify}} + "#" + {{decl.var.stringify}} + " cannot be nil" if @{{decl.var}}.nil?
-        @{{decl.var}}.not_nil!
-      end
-    {% else %}
-      def {{decl.var.id}}=(@{{decl.var.id}} : {{type.id}}); end
-
-      def {{decl.var.id}} : {{type.id}}
-        raise NilAssertionError.new {{@type.name.stringify}} + "#" + {{decl.var.stringify}} + " cannot be nil" if @{{decl.var}}.nil?
-        @{{decl.var}}.not_nil!
-      end
-    {% end %}
+    {% if !primary || (primary && !auto) %} property{{(nilable || !decl.value.is_a?(Nop) ? "" : '!').id}} {% else %} getter {% end %} {{decl.var}} : {{decl.type}} {% unless decl.value.is_a? Nop %} = {{decl.value}} {% end %}
```

Comment:
since all the columns are defined as properties (accept id), can we just use the default initializer?

Reply:
Property macro doesn't create an initializer, if we go with option 3, then the user could define whatever they want it to be.

Reply:
I am thinking that we don't create an initializer and leave it up to the user how to set the properties.  This will make the class less magic and will handle the `virtual` fields.  They can use `getter` or `property` macros or override the getter and setter methods manually.  Leave it up to the user how to set the properties.

Reply:
That is one of the benefits.  However I'll leave this open for a while as I'd rather have other maintainers/users give their thoughts as well.